mlir/unittests/pipeline/test_compiler_pipeline.cpp

```diff
@@ -0,0 +1,1379 @@
+// # Temporary Unitary Operation Tests
+// ##################################################
+
+TEST_F(CompilerPipelineTest, X) {
+  qc::QuantumComputation qc;
+  qc.addQubitRegister(1, "q");
+  qc.x(0);
+  qc.x(0);
+  qc.x(0);
+
+  const auto module = importQuantumCircuit(qc);
+  ASSERT_TRUE(module);
+  ASSERT_TRUE(runPipeline(module.get()).succeeded());
+
+  const auto quartzInit = buildQuartzIR([](quartz::QuartzProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(1, "q");
+    auto q = reg[0];
+    b.x(q);
+    b.x(q);
+    b.x(q);
+  });
+  const auto fluxInit = buildFluxIR([](flux::FluxProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(1, "q");
+    auto q0 = reg[0];
+    auto q1 = b.x(q0);
+    auto q2 = b.x(q1);
+    b.x(q2);
+  });
+  const auto fluxOpt = buildFluxIR([](flux::FluxProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(1, "q");
+    auto q0 = reg[0];
+    b.x(q0);
+  });
+  const auto quartzOpt = buildQuartzIR([](quartz::QuartzProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(1, "q");
+    auto q = reg[0];
+    b.x(q);
+  });
+  const auto qirOpt = buildQIR([](qir::QIRProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(1);
+    auto q = reg[0];
+    b.x(q);
+  });
+
+  verifyAllStages({
+      .quartzImport = quartzInit.get(),
+      .fluxConversion = fluxInit.get(),
+      .optimization = fluxOpt.get(),
+      .quartzConversion = quartzOpt.get(),
+      .qirConversion = qirOpt.get(),
+  });
+}
+
+TEST_F(CompilerPipelineTest, CX) {
+  qc::QuantumComputation qc;
+  qc.addQubitRegister(2, "q");
+  qc.cx(0, 1);
+
+  const auto module = importQuantumCircuit(qc);
+  ASSERT_TRUE(module);
+  ASSERT_TRUE(runPipeline(module.get()).succeeded());
+
+  const auto quartzInit = buildQuartzIR([](quartz::QuartzProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(2, "q");
+    auto q0 = reg[0];
+    auto q1 = reg[1];
+    b.ctrl(q0, [&](auto& b) { b.x(q1); });
+  });
+  const auto fluxInit = buildFluxIR([](flux::FluxProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(2, "q");
+    auto q0 = reg[0];
+    auto q1 = reg[1];
+    b.ctrl({q0}, {q1}, [&](auto& b) {
+      auto q1res = b.x(q1);
+      return SmallVector<Value>{q1res};
+    });
+  });
+  const auto fluxOpt = buildFluxIR([](flux::FluxProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(2, "q");
+    auto q0 = reg[0];
+    auto q1 = reg[1];
+    b.ctrl({q0}, {q1}, [&](auto& b) {
+      auto q1res = b.x(q1);
+      return SmallVector<Value>{q1res};
+    });
+  });
+  const auto quartzOpt = buildQuartzIR([](quartz::QuartzProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(2, "q");
+    auto q0 = reg[0];
+    auto q1 = reg[1];
+    b.ctrl(q0, [&](auto& b) { b.x(q1); });
+  });
+  const auto qirOpt = buildQIR([](qir::QIRProgramBuilder& b) {
+    auto reg = b.allocQubitRegister(2);
+    auto q0 = reg[0];
+    /// TODO: Replace uncomment this when CX can be converted to QIR
+    // auto q1 = reg[1];
+    // b.cx(q0, q1);
+    b.x(q0);
```

Comment:
Once the conversion is implemented correctly, the test should pass with this:

```suggestion
    auto q1 = reg[1];
    b.cx(q0, q1);
```